### DISCUSSION TITLE
How to compute all satisfying models/solutions?

### DISCUSSION THREAD

**Discussion by sin2y** [idrave/FOND-ASP#39] [Q&A]
**Description:**  
As far as I understand, ASP can be used to find all satisfying models. I’ve made some modifications to the code to enable this, but I’m not certain if these changes are correct. I’d appreciate any guidance on whether my approach is valid, or any suggestions on how to correctly configure the code to obtain all satisfying models.

**Questions:**
- Could anyone review my modifications to confirm if they effectively capture all models?
- Are there recommended settings or parameters to ensure that all satisfying models are found?

**My Modifications:**  
1. Modify file `fondplus_show_pretty.py`
```python
# file: fcfond/planner_clingo/fondplus_show_pretty.py
...
107 def main(prg):
108     # just search for one model and that's it
109 -    main_one(prg)
109 +    # main_one(prg)
...
117 -     # main_many(prg)
117 +    main_many(prg)
```
2. run commands as usual, like
```shell
python -m fcfond.main  -pddl domains/pddl/synthetic/qnp2/domain03.pddl domains/pddl/synthetic/qnp2/p03.pddl -log -stats
```
Is this right?

**Comment by idrave** [comment]
@ArcYuan your changes should indeed lead to all solutions, but I would recommend running the `fcfond.main` file with option `-n` set to 0, without requiring any code changes. When you specify the number of models (solutions) to be found to clingo, 0 represents searching for all solutions, and the option `-n` is forwarded to clingo in this way. The `main_many` function essentially overrides this parameter to be 0.

**Reply by sin2y** [reply]
Thank you for the suggestion! However, when testing the `-n 0` parameter in my setup, I noticed that adding the `-n 0` option may not always achieve the intended outcome in some cases. Here is a detailed summary of my experiments and environment settings:

### Experiment Environment
- **Operating System**: Ubuntu 20.04 x86_64
- **Python Version**: Python 3.9, installed with `conda create -n fond-asp python=3.9`
- **Clingo Version**: Clingo 5.5.1 installed with `conda install -c potassco clingo=5.5.1`

The entire setup process is as follows:
```shell
conda create -n fond-asp python=3.9
conda activate fond-asp
conda install -c potassco clingo=5.5.1
git clone https://github.com/idrave/fond-asp.git
cd fond-asp/
pip install -r requirements.txt
pip install -e . 
```

### Test Results

1. **Example 1 (Successful): For strong-cyclic planning**

Command:
```shell
python -m fcfond.main -stats -pddl domains/pddl/synthetic/qnp2/domain05.pddl domains/pddl/synthetic/qnp2/p05.pddl --strongcyclic -n 0
```
Output:
```shell
...
ASP Solved. Processing output
Output processed.
Problem: p05
States: 40
Actions: 6
Pre-processing time: 0.022
Sat: True
Models: 345
Calls: 1
Time: 0.012
Solve Time: 0.01
1st Model Time: 0.0
Unsat Time: 0.0
CPU Time: 0.012
Result: True
Max Memory: 42.788

  Problem  States  Actions  Pre-processing time   Sat Models  Calls   Time  Solve Time  1st Model Time  Unsat Time  CPU Time Result  Max Memory
0     p05      40        6                0.022  True    345      1  0.012        0.01             0.0         0.0     0.012   True      42.788
```
    
Works well for finding all strong-cyclic solutions without code modifications.

2. **Example 2 (Unsuccessful)**: For FOND+ planning

Command:
```shell
python -m fcfond.main -stats -pddl domains/pddl/synthetic/qnp2/domain05.pddl domains/pddl/synthetic/qnp2/p05.pddl -n 0
```
Output:
```shell
...
ASP Solved. Processing output
Output processed.
Problem: p05
States: 40
Actions: 6
Pre-processing time: 0.022
Sat: True
Models: 1+
Calls: 1
Time: 0.064
Solve Time: 0.0
1st Model Time: 0.0
Unsat Time: 0.0
CPU Time: 0.064
Result: True
Max Memory: 42.804

  Problem  States  Actions  Pre-processing time   Sat Models  Calls   Time  Solve Time  1st Model Time  Unsat Time  CPU Time Result  Max Memory
0     p05      40        6                0.022  True     1+      1  0.064         0.0             0.0         0.0     0.064   True      42.804
```
**Fails** to find all models.

### Request
Given these results, could you please clarify if there are any known compatibility issues or configurations that might affect the behavior of the `-n 0` option? Are there any specific Clingo and Python versions you recommend for this?

Thank you again for your assistance!
